pkg/cluster/cluster.go

```diff
@@ -0,0 +1,432 @@
+	 */
+	if !cluster.CreateClusterOpts.DisableImageVolume {
+		imageVolumeName := fmt.Sprintf("%s-%s-images", k3d.DefaultObjectNamePrefix, cluster.Name)
+		if err := runtime.CreateVolume(imageVolumeName, map[string]string{"k3d.cluster": cluster.Name}); err != nil {
+			log.Errorln("Failed to create image volume '%s' for cluster '%s'", imageVolumeName, cluster.Name)
+			return err
+		}
+
+		extraLabels["k3d.cluster.volumes.imagevolume"] = imageVolumeName
+
+		// attach volume to nodes
+		for _, node := range cluster.Nodes {
+			node.Volumes = append(node.Volumes, fmt.Sprintf("%s:%s", imageVolumeName, k3d.DefaultImageVolumeMountPath))
+		}
+	}
+
+	/*
+	 * Nodes
+	 */
+
+	nodeSetup := func(node *k3d.Node, suffix int) error {
+		// cluster specific settings
+		if node.Labels == nil {
+			node.Labels = make(map[string]string) // TODO: maybe create an init function?
+		}
+		node.Labels["k3d.cluster"] = cluster.Name
+		node.Env = append(node.Env, fmt.Sprintf("K3S_TOKEN=%s", cluster.Secret))
+		node.Labels["k3d.cluster.secret"] = cluster.Secret
+
+		// append extra labels
+		for k, v := range extraLabels {
+			node.Labels[k] = v
+		}
+
+		// node role specific settings
+		if node.Role == k3d.MasterRole {
+
+			// the cluster has an init master node, but its not this one, so connect it to the init node
+			if cluster.InitNode != nil && !node.MasterOpts.IsInit {
+				node.Args = append(node.Args, "--server", fmt.Sprintf("https://%s:%d", cluster.InitNode.Name, 6443))
+			}
+
+		} else if node.Role == k3d.WorkerRole {
+			// connection url
+			connectionURL := fmt.Sprintf("https://%s:%d", generateNodeName(cluster.Name, k3d.MasterRole, 0), 6443)
+			node.Env = append(node.Env, fmt.Sprintf("K3S_URL=%s", connectionURL))
+			node.Labels["k3d.cluster.url"] = connectionURL
+		}
+
+		node.Name = generateNodeName(cluster.Name, node.Role, suffix)
+		node.Network = cluster.Network.Name
+
+		// create node
+		log.Infof("Creating node '%s'", node.Name)
+		if err := CreateNode(node, runtime); err != nil {
+			log.Errorln("Failed to create node")
+			return err
+		}
+		log.Debugf("Created node '%s'", node.Name)
+
+		return err
+	}
+
+	// used for node suffices
+	masterCount := 0
+	workerCount := 0
+	suffix := 0
+
+	// create init node first
+	if cluster.InitNode != nil {
+		log.Infoln("Creating initializing master node")
+		cluster.InitNode.Args = append(cluster.InitNode.Args, "--cluster-init")
+		if err := nodeSetup(cluster.InitNode, masterCount); err != nil {
+			return err
+		}
+		masterCount++
+
+		// wait for the initnode to come up before doing anything else
+		for {
+			log.Debugln("Waiting for initializing master node...")
+			logreader, err := runtime.GetNodeLogs(cluster.InitNode)
+			defer logreader.Close()
+			if err != nil {
+				logreader.Close()
+				log.Errorln(err)
+				log.Errorln("Failed to get logs from the initializig master node.. waiting for 3 seconds instead")
+				time.Sleep(3 * time.Second)
+				goto initNodeFinished
```

Comment:
can this just be break.  goto should be avoided at all costs.

Reply:
Definitely :+1: